From 88203f65cdd92a56177d40519673689c05c47d2f Mon Sep 17 00:00:00 2001
From: Jason Wang <jasowang@redhat.com>
Date: Thu, 9 Oct 2014 06:13:30 -0400
Subject: [virt] tools/hyperv: Handle the case when the target file exists correctly

Message-id: <1412835210-22435-12-git-send-email-jasowang@redhat.com>
Patchwork-id: 97293
O-Subject: [RHEL7.1 PATCH V2 11/11] Tools: hv: Handle the case when the target file exists correctly
Bugzilla: 1146358
RH-Acked-by: Vitaly Kuznetsov <vkuznets@redhat.com>

From: K. Y. Srinivasan <kys@microsoft.com>

Bugzilla: 1146358

Return the appropriate error code and handle the case when the target
file exists correctly. This fixes a bug.

Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Cc: <stable@vger.kernel.org> [3.14]
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
(cherry picked from commit 314672a2c2780212fb770bb02d2fffaa1019823f)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/uapi/linux/hyperv.h b/include/uapi/linux/hyperv.h
index 9beb7c9..78e4a86 100644
--- a/include/uapi/linux/hyperv.h
+++ b/include/uapi/linux/hyperv.h
@@ -305,6 +305,7 @@ enum hv_kvp_exchg_pool {
 #define HV_ERROR_DEVICE_NOT_CONNECTED 0x8007048F
 #define HV_INVALIDARG   0x80070057
 #define HV_GUID_NOTFOUND  0x80041002
+#define HV_ERROR_ALREADY_EXISTS  0x80070050
 
 #define ADDR_FAMILY_NONE 0x00
 #define ADDR_FAMILY_IPV4 0x01
diff --git a/tools/hv/hv_fcopy_daemon.c b/tools/hv/hv_fcopy_daemon.c
index 4ecc4fd..fba1c75 100644
--- a/tools/hv/hv_fcopy_daemon.c
+++ b/tools/hv/hv_fcopy_daemon.c
@@ -82,8 +82,10 @@ static int hv_start_fcopy(struct hv_start_fcopy *smsg)
 
  if (!access(target_fname, F_OK)) {
   syslog(LOG_INFO, "File: %s exists", target_fname);
-  if (!smsg->copy_flags & OVER_WRITE)
+  if (!(smsg->copy_flags & OVER_WRITE)) {
+   error = HV_ERROR_ALREADY_EXISTS;
    goto done;
+  }
  }
 
  target_fd = open(target_fname, O_RDWR | O_CREAT | O_CLOEXEC, 0744);
-- 
1.7.1